### PR TITLE
Condition of sale

### PR DESCRIPTION
**Pull request by matti-lamppu**
# Hitas Pull Request

# Description

Adds the "condition of sale" model and endpoints.

> Does not yet add the ability to filter apartments based on if they have conditions of sale of not. 

## Pull request checklist

Check the boxes for each DoD items that has been completed:

- **Frontend**
    - [ ] Changes have been tested
- **Backend**
    - [x] Changes have been tested
    - [x] Automatic tests have been added
    - [x] OpenAPI definitions have been updated
    - [x] Database migrations will work in the test environment
    - [ ] Oracle migration has been updated
    - [x] initial.json work with current migrations

## Test plan

- [x] Automatic tests
- [ ] Check logic for edge cases

## Tickets

This pull request resolves all or part of the following ticket(s): HT-173, HT-175

### CODE REVIEW

File: backend/hitas/views/condition_of_sale.py [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+        # We have to fetch ownerships separately, since if only some conditions of sale in 'to_save' were created,
+        # the ids in the returned list from 'bulk_create' are not correct.
+        conditions_of_sale: list[ConditionOfSale] = list(
+            ConditionOfSale.objects.select_related("new_ownership__owner", "old_ownership__owner")
+            .filter(Q(new_ownership__owner=owner) | Q(old_ownership__owner=owner))
+            .all()
+        )
+
+        validated_data["conditions_of_sale"] = conditions_of_sale
+        return validated_data["conditions_of_sale"]
```
Comment (ranta) [inline backend/hitas/views/condition_of_sale.py:137]: Why set the value to the dict, then immediately return it 🤔

File: backend/hitas/views/condition_of_sale.py [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+                        "purchase_date",
+                    ).order_by("purchase_date"),
+                ),
+            ).get(uuid=owner_uuid)
+        except Owner.DoesNotExist as error:
+            raise HitasModelNotFound(model=Owner) from error
+
+        return attrs
+
+    def create(self, validated_data: dict[str, Any]) -> list[ConditionOfSale]:
```
Comment (ranta) [inline backend/hitas/views/condition_of_sale.py:105]: This method could use a comment about what/why is happening, as its a pretty irregular model endpoint

**Review by ranta — CHANGES_REQUESTED** [review]
Let's do smaller PRs in the future, please 🙂

File: backend/hitas/tests/factories/test_factories.py [hunk 3]
```diff
@@ -21,9 +21,17 @@
...
     assert factory_class._meta.model.objects.count() == 1
+
+
+@pytest.mark.django_db
+def test__factory__condition_of_sale_factory():
+    owner = factories.OwnerFactory.create()
+    ownership_1 = factories.OwnershipFactory.create(owner=owner)
+    ownership_2 = factories.OwnershipFactory.create(owner=owner)
+    factories.ConditionOfSaleFactory.create(new_ownership=ownership_1, old_ownership=ownership_2)
+    assert factories.ConditionOfSaleFactory._meta.model.objects.count() == 1
```
[the remark this answers is not in the document]
Comment (matti-lamppu) [reply backend/hitas/tests/factories/test_factories.py:37]: Nice, made the change

File: backend/hitas/views/condition_of_sale.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+class MinimalApartmentSerializer(HitasModelSerializer):
+    class Meta:
+        model = Apartment
+        fields = [
+            "id",
+            "street_address",
+            "apartment_number",
+            "floor",
+            "stair",
+        ]
```
[the remark this answers is not in the document]
Comment (matti-lamppu) [reply backend/hitas/views/condition_of_sale.py:24]: Both `hitas.views.apartment_list.ApartmentListSerializer` and `hitas.views.apartment.ApartmentHitasAddressSerializer` have fields we don't need in the response for a condition of sale, or are missing some of the fields. We could create an inheritance hierarchy to not duplicate the setup, but since there isn't any business logic in the serializer, I think we can just leave it as is.

File: backend/hitas/models/condition_of_sale.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+        related_name="conditions_of_sale_new",
+        on_delete=models.CASCADE,
+        editable=False,
+    )
+    old_ownership = models.ForeignKey(
+        "Ownership",
+        related_name="conditions_of_sale_old",
+        on_delete=models.CASCADE,
+        editable=False,
+    )
```
[the remark this answers is not in the document]
Comment (matti-lamppu) [reply backend/hitas/models/condition_of_sale.py:38]: I think in this case the ownership of the old apartment should be corrected so that it reflects the actual owners of that apartment. Then we can create the condition of sale for the correct ownerships, since Person A does not actually have to give up their share of the ownership, only Person B's share, which is would be falsely included in Person A's share.

[the remark this answers is not in the document]
Comment (ranta) [reply backend/hitas/models/condition_of_sale.py:38]: In my example **Person A** does need to give up their share of the old apartment, because **Person A** and **B** are in the same household (and also will be in the future).
In this example **Person B** also does not own any shares in the old apartment, and **Person A** will not be owning any shares of the new apartment.

[the remark this answers is not in the document]
Comment (matti-lamppu) [reply backend/hitas/models/condition_of_sale.py:38]: I thought that people sharing a household would implicitly share the shares for that apartment? That is how we have been talking about it with the customer at least, that the "other person needs to give up their share".

In a shared household this will usually require the other person to also I've up their share, since they likely want to continue sharing the household, but that is just one possible arrangement between the owners.

[the remark this answers is not in the document]
Comment (ranta) [reply backend/hitas/models/condition_of_sale.py:38]: > would implicitly share the shares for that apartment
No, owning shares is personal. The need for the other person to give up shares comes from being in the household.

Two people owning shares in an apartment does not mean they belong in the same household, and also two people can belong in the same household without both of them owning shares in the apartment they live in.

[the remark this answers is not in the document]
Comment (matti-lamppu) [reply backend/hitas/models/condition_of_sale.py:38]: Ok, then we should remove the restriction. To support these cases, I think we should include an optional field in the condition of sale create endpoint under owner where you can add additional ownerships that are for new apartments the owner doesn't have ownerships in. We can then create additional conditions of sale between these ownerships and the persons current ownership.

[the remark this answers is not in the document]
Comment (ranta) [reply backend/hitas/models/condition_of_sale.py:38]: Or, if we replace `new_ownership` with `new_apartment` we remove one more 'duplicate' field (`ownership.owner`) and remove the need to add a public id to `Ownership`, while keeping every request the same (no need for optional fields) 😉

File: backend/hitas/models/condition_of_sale.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+
+    def clean(self) -> None:
+        if self.new_ownership.owner != self.old_ownership.owner:
+            raise ValidationError("Owners must be the same.")
+
+    def __str__(self) -> str:
+        return f"{self.new_ownership.owner}: {self.new_ownership.apartment} -> {self.old_ownership.apartment}"
+
+
+def condition_of_sale_queryset() -> models.QuerySet[ConditionOfSale]:
```
[the remark this answers is not in the document]
Comment (matti-lamppu) [reply backend/hitas/models/condition_of_sale.py:73]: SafeDeleteQuerySet already creates three model managers, and we would need to override all of them, and any subsequent managers we might add in the future.

IMO, we should have a service layer where we would have our database queries reusable between views. A separate service layer independent of Django also gives us the freedom to change things easier in the future. This is a bigger change and should be discussed in the architecture review.

File: backend/hitas/utils.py [hunk 7]
```diff
@@ -78,3 +79,18 @@ def lookup_model_id_by_uuid(lookup_id: str, model_class: type[models.Model], **k
...
+def get_many_or_cached_prefetch(
+    instance: models.Model,
+    field_name: str,
+    # Optional argument to provide return type hints
+    related_model: type[TModel] = ModelBase,  # noqa
+) -> models.QuerySet[TModel]:
+    if hasattr(instance, "_prefetched_objects_cache") and field_name in instance._prefetched_objects_cache:
+        return instance._prefetched_objects_cache[field_name]
+    else:
+        return getattr(instance, field_name).all()
```
[the remark this answers is not in the document]
Comment (matti-lamppu) [reply backend/hitas/utils.py:96]: This function only matters when the prefetched queryset has deferred fields, and we need to access the prefetched items, e.g., in the loop. Ownerships are accessed from `apartment.ownerships`, but this is a ModelManager class that cannot be iterated, so we need to evaluate a queryset from it. Usually, you would do `apartment.ownerships.all()`, but this will now create another query since the prefetched values had deferred fields, and therefore the queryset in the prefetch cache cannot be used. This function uses the prefetched queryset if one is found and then uses `.all()` if the cache is not found. If the prefetch cache has cached a queryset matching `all()`, then prefetch cache can be used, and the number of queries is the same, and thus this function is unnecessary, as in your example.

File: backend/hitas/views/condition_of_sale.py (hunk 1 above)
Comment (matti-lamppu) [reply backend/hitas/views/condition_of_sale.py:137]: No reason, I'll simplify this.

File: backend/hitas/models/condition_of_sale.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (matti-lamppu) [reply backend/hitas/models/condition_of_sale.py:38]: Using just an apartment reference will make canceling the sale (or removing the ownership for the new apartment some other way if we decide so in the future) for the new apartment not automatically fulfill its conditions of sale, so I would prefer using two ownership references and leave it to cascade.

File: backend/hitas/models/apartment.py [hunk 8]
```diff
@@ -109,6 +114,20 @@ def old_hitas_ruleset(self) -> bool:
...
+    def is_new(self) -> bool:
+        today = timezone.now().date()
+        if self.completion_date is None or self.completion_date > today:
+            return True
+
+        sales: list["ApartmentSale"] = list(get_many_or_cached_prefetch(self, "sales"))
+        no_sales = len(sales) < 1
+
+        no_first_purchase_or_in_the_future = self.first_purchase_date is None or self.first_purchase_date > today
+        no_first_sale_or_in_the_future = no_sales or sales[0].purchase_date > today
```
[the remark this answers is not in the document]
Comment (matti-lamppu) [reply backend/hitas/models/apartment.py:127]: If we decide to migrate the sales history from the old system, we can remove the `first_purchase_date` from apartment, as that is then the information of the sale and not the apartment, and only leave the other check. But before that, we still use the apartment field to keep track of this, so that check is also needed. Otherwise, all apartments would be considered new, which would complicate testing right now.

File: backend/hitas/views/condition_of_sale.py (hunk 2 above)
Comment (matti-lamppu) [reply backend/hitas/views/condition_of_sale.py:105]: Ok, added some explanation to this and the validate method.

File: backend/hitas/views/utils/viewsets.py [hunk 9]
```diff
@@ -1,42 +1,102 @@
...
 from django.http import Http404
-from rest_framework import viewsets
+from rest_framework import serializers, viewsets
 
 from hitas.exceptions import HitasModelNotFound
 from hitas.utils import lookup_id_to_uuid
 from hitas.views.utils import HitasPagination
 
 
 class HitasModelMixin:
```
[the remark this answers is not in the document]
Comment (matti-lamppu) [reply backend/hitas/views/utils/viewsets.py:10]: It's starting to look like library code on top of drf, which I think is fine. Personally, I don't find it too bad (although it's me who wrote it), and I was able to remove some of the `lookup_id_to_uuid` previously, which helps.

File: backend/hitas/utils.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (matti-lamppu) [reply backend/hitas/utils.py:96]: Actually, I tested this and looked in a bit deeper into the django codebase, and it seem like [`.all()` does actually use the prefetched queryset](https://github.com/django/django/blob/main/django/db/models/fields/related_descriptors.py#L1082). In my testing last week it didn't work for some reason, but I may have been simply missing some fields in the `.only()` definition. I'll remove this method and replace them with `.all()` calls.

File: backend/hitas/models/apartment.py [hunk 10]
```diff
@@ -109,6 +114,20 @@ def old_hitas_ruleset(self) -> bool:
     def address(self) -> str:
         return f"{self.street_address} {self.stair} {self.apartment_number}"
 
+    @property
+    def is_new(self) -> bool:
+        today = timezone.now().date()
+        if self.completion_date is None or self.completion_date > today:
+            return True
+
+        sales: list["ApartmentSale"] = list(get_many_or_cached_prefetch(self, "sales"))
```
[the remark this answers is not in the document]
Comment (matti-lamppu) [reply backend/hitas/models/apartment.py:123]: Doesn't seem like Prefetch supports limits (via slicing) but I can make it work with a silly subquery trick..

File: backend/hitas/migrations/0051_add_condition_of_sale.py [hunk 11]
```diff
@@ -1,4 +1,4 @@
-# Generated by Django 3.2.16 on 2023-01-18 18:47
```
Comment (ranta) [inline backend/hitas/migrations/0051_add_condition_of_sale.py]: > Remove ownership uuid

Instead of adding it and removing it, can you instead delete the original commit that added it?

Comment (matti-lamppu) [reply backend/hitas/migrations/0051_add_condition_of_sale.py]: I think I have 10+ commits on this branch between when I added the Onwership UUID and removing it, I would rather not go through the rebase hell of rewriting all those commits. I can do a squash at the end if we want a cleaner history.

File: backend/hitas/views/ownership.py [hunk 12]
```diff
@@ -25,12 +26,18 @@ class Meta:
...
     owner = OwnerSerializer()
     percentage = HitasDecimalField(required=True)
 
+    def get_id(self, instance: ExternalHitasModel) -> str:
+        return instance.uuid.hex
+
     class Meta:
         model = Ownership
         fields = [
+            "id",
```
Comment (ranta) [inline backend/hitas/views/ownership.py:40]: `id` can be removed from here as well :)

File: backend/hitas/tests/apis/test_api_condition_of_sale.py [hunk 13]
```diff
@@ -0,0 +1,1179 @@
...
+    )
+    response = api_client.get(url)
+    assert response.status_code == status.HTTP_200_OK, response.json()
+    assert response.json()["contents"] == [
+        {
+            "id": condition_of_sale.uuid.hex,
+            "grace_period": str(condition_of_sale.grace_period.value),
+            "fulfilled": condition_of_sale.fulfilled,
+            "new_ownership": {
+                "id": new_ownership.uuid.hex,
```
Comment (ranta) [inline backend/hitas/tests/apis/test_api_condition_of_sale.py:74]: And here

File: backend/hitas/models/condition_of_sale.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (matti-lamppu) [reply backend/hitas/models/condition_of_sale.py:37]: Added the option to add additional ownerships for which to create conditions of sale. This brings back the ownership UUID so that it can be used to set the additional ownerships.

File: backend/hitas/views/ownership.py (hunk 12 above)
Comment (matti-lamppu) [reply backend/hitas/views/ownership.py:40]: Additional ownerships option on condition of sale creation needs this, see [here](https://github.com/City-of-Helsinki/hitas/pull/212/files#diff-371f597cf9397b67f5f3749ab0f5c08c290e53cf39589034af23ead33edee12bR79).

File: backend/hitas/tests/apis/test_api_condition_of_sale.py (hunk 13 above)
Comment (matti-lamppu) [reply backend/hitas/tests/apis/test_api_condition_of_sale.py:74]: Additional ownerships option on condition of sale creation needs this, see [here](https://github.com/City-of-Helsinki/hitas/pull/212/files#diff-371f597cf9397b67f5f3749ab0f5c08c290e53cf39589034af23ead33edee12bR79).

File: backend/hitas/migrations/0051_ownership_uuid.py [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+        # From django docs:
+        # https://docs.djangoproject.com/en/3.2/howto/writing-migrations/#migrations-that-add-unique-fields
+        #
+        # Note there is a race condition if we allow objects to be created while this migration is running.
+        # Objects created after the AddField and before RunPython will have their original uuid’s overwritten.
+        migrations.AddField(
+            model_name="ownership",
+            name="uuid",
+            field=models.UUIDField(default=uuid.uuid4),
+        ),
```
[the remark this answers is not in the document]
Comment (matti-lamppu) [reply backend/hitas/migrations/0051_ownership_uuid.py]: Ownership UUID is now removed since the new API uses households with owner UUIDs.

File: backend/hitas/models/condition_of_sale.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (matti-lamppu) [reply backend/hitas/models/condition_of_sale.py:37]: Ownership UUID is now removed since the new API uses households with owner UUIDs.

**Comment by matti-lamppu** [comment]
New condition of sale create endpoint with `household` instead of `additional_ownerships` is now available. Tests have also been written with [given-when-then](https://en.wikipedia.org/wiki/Given-When-Then) to make them easier to read.

**Review by ranta — APPROVED** [review]
Nice work 🚀
